### PR TITLE
Move DataTable sink for CSV/DataStreams to DataTables

### PR DESCRIPTION
**Pull request by cjprybol**

### CODE REVIEW

File: REQUIRE [hunk 1]
```diff
@@ -5,3 +5,6 @@ StatsBase 0.11.0
 SortingAlgorithms
 Reexport
 Compat 0.19.0
+WeakRefStrings
+CSV
```
Comment (nalimilan) [inline REQUIRE]: Is this actually needed?

File: src/DataTables.jl [hunk 2]
```diff
@@ -17,7 +17,7 @@ import NullableArrays: dropnull, dropnull!
 @reexport using CategoricalArrays
 using GZip
 using SortingAlgorithms
-
+using WeakRefStrings
```
Comment (nalimilan) [inline src/DataTables.jl]: Can you move this to `io.jl` so that it's clear where it's used?

File: src/abstractdatatable/io.jl [hunk 3]
```diff
@@ -193,3 +193,122 @@ end
 @compat function Base.show(io::IO, ::MIME"text/tab-separated-values", dt::AbstractDataTable)
     printtable(io, dt, true, '\t')
 end
+
+##############################################################################
+#
+# CSV IO
```
Comment (nalimilan) [inline src/abstractdatatable/io.jl]: Not only CSV, DataStreams.

File: src/abstractdatatable/io.jl [hunk 4]
```diff
@@ -193,3 +193,122 @@ end
...
+# DataTable as a Data.Source
+function Data.isdone(source::DataTable, row, col)
+    rows, cols = size(source)
+    return row > rows || col > cols
+end
+
+Data.streamtype(::Type{DataTable}, ::Type{Data.Column}) = true
+Data.streamtype(::Type{DataTable}, ::Type{Data.Field}) = true
+
+Data.streamfrom{T <: AbstractVector}(source::DataTable, ::Type{Data.Column}, ::Type{T}, col) = (@inbounds A = source.columns[col]::T; return A)
```
Comment (nalimilan) [inline src/abstractdatatable/io.jl]: Please wrap long lines.

File: src/abstractdatatable/io.jl [hunk 5]
```diff
@@ -193,3 +193,122 @@ end
...
+# CSV IO
+#
+##############################################################################
+
+importall DataStreams
+
+# DataTables DataStreams implementation
+function Data.schema(df::DataTable, ::Type{Data.Column})
+    return Data.Schema(map(string, names(df)),
+            DataType[typeof(A) for A in df.columns], size(df, 1))
```
Comment (nalimilan) [inline src/abstractdatatable/io.jl]: Incorrect alignment.

File: src/abstractdatatable/io.jl [hunk 6]
```diff
@@ -193,3 +193,122 @@ end
...
+allocate{T}(::Type{Nullable{T}}, rows, ref) = NullableArray{T, 1}(Array{T}(rows), fill(true, rows), isempty(ref) ? UInt8[] : ref)
+allocate{T}(::Type{NullableVector{T}}, rows, ref) = NullableArray{T, 1}(Array{T}(rows), fill(true, rows), isempty(ref) ? UInt8[] : ref)
+
+allocate{S,R}(::Type{CategoricalArrays.CategoricalValue{S,R}}, rows, ref) = CategoricalArray{S,1,R}(rows)
+allocate{S,R}(::Type{CategoricalVector{S,R}}, rows, ref) = CategoricalArray{S,1,R}(rows)
+
+allocate{S,R}(::Type{Nullable{CategoricalArrays.CategoricalValue{S,R}}}, rows, ref) = NullableCategoricalArray{S,1,R}(rows)
+allocate{S,R}(::Type{NullableCategoricalVector{S,R}}, rows, ref) = NullableCategoricalArray{S,1,R}(rows)
+
+if isdefined(Main, :DataArray)
```
Comment (nalimilan) [inline src/abstractdatatable/io.jl]: I guess this won't play well with precompilation?

**Comment by cjprybol** [comment]
These changes alone would be sufficient to support DataTables without making any changes in DataStreams, CSV, or DataFrames, right? I can use `CSV.read(file, DataTable, args...)` to read into a DataTable with this while using master/release branches of the other 3, although I'm getting segfaults when I try and work with the WeakRefStrings. I'll need to figure that out, and we'd need to fix [this](https://github.com/quinnj/WeakRefStrings.jl/issues/5) too. We could drop the functions that don't have anything DataTables specific, which would be the [`allocate`](https://github.com/JuliaData/DataTables.jl/pull/35/files#diff-93e22f56a07e6e0568a6dcf1a75e4b8cR229) functions and WeakRefString-specific [`append!`](https://github.com/JuliaData/DataTables.jl/pull/35/files#diff-93e22f56a07e6e0568a6dcf1a75e4b8cR316) and bigger decisions about whether functionality should move to DataTables & DataFrames, or directly to AbstractTables with DataTables and DataFrames inheriting from AbstractTables could be made later?

**Comment by quinnj** [comment]
FYI, definitely still on board w/ the DataStreams refactoring, I'm just trying to get new tags out for packages before merging all the refactorings (that way they all have at least one tag that works for 0.6 before more major upheaval).

**Comment by quinnj** [comment]
Pulled this locally to try out and kick the tires. It's actually pretty straightforward, just porting the code over from DataStreams and updating for DataTables. It's nice too because this doesn't require any changes to DataStreams or DataFrames. I did throw in a quick test and updates the REQUIRES and test/REQUIRES.

**Comment by cjprybol** [comment]
Thanks Jacob!

**Pull request merged**